persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -113,11 +211,80 @@ runConfiguredParser acc parser fp s = parseResult
 renderErrors :: [EntityParseError] -> String
 renderErrors errs = intercalate "\n" $ fmap errorBundlePretty errs
 
-toCumulativeParseResult
-    :: (Monoid a) => [ParseResult a] -> CumulativeParseResult a
-toCumulativeParseResult prs = case partitionEithers prs of
-    ([], results) -> Right $ fold results
-    (errs, _) -> Left errs
+-- | Renders a list of ParserWarnings as a String,
+-- separated by line breaks.
+-- @since 2.16.0.0
+renderWarnings :: [ParserWarning] -> String
+renderWarnings warnings = intercalate "\n" $ fmap parserWarningMessage warnings
+
+-- Attempts to parse with a provided parser. If it fails with an error matching
+-- the provided predicate, it registers a warning with the provided message and falls
+-- back to the second provided parser.
+tryOrWarn
+    :: String
+    -> (ParseError String Void -> Bool)
+    -> Parser a
+    -> Parser a
+    -> Parser a
+tryOrWarn msg f l r = do
+    parserState <- getParserState
+    withRecovery (warnAndRetry $ statePosState parserState) l
+  where
+    warnAndRetry posState err = do
+        if f err
+            then do
+                let
+                    (pairs, _) = attachSourcePos errorOffset [err] posState
+                tell $
+                    map
+                        ( \(e, _pos) ->
+                            ParserWarning
+                                { parserWarningExtraMessage = msg <> "\n"
+                                , parserWarningUnderlyingError = e
+                                , parserWarningPosState = posState
+                                }
+                        )
+                        pairs
+                r
+            else parseError err
+
+--  Attempts to parse with a provided parser. If it fails with an error matching
+-- the provided predicate, it registers a delayed error with the provided message and falls
+-- back to the second provided parser.
+--
+-- This is useful when registering errors in space consumers and other parsers that are called
+-- with `try`, since a non-delayed error in this context will cause backtracking and not
+-- get reported to the user.
+tryOrRegisterError
+    :: String
+    -> (ParseError String Void -> Bool)
+    -> Parser a
+    -> Parser a
+    -> Parser a
+tryOrRegisterError msg f l r = do
```

Comment:
If using single variable names, it is idiomatic to use `p` with a predicate (ie `a -> Bool`)